[ONNX] Initial version of diagnostics infrastructure.

BowenBao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #85107
* #85918
* #85898
* #85651
* #85428

This PR introduces a general Python diagnostics infrastructure powered by SARIF,
and the exporter diagnostics module that builds on top of it.

BowenBao committed:
[ONNX][WIP] Initial version of diagnostics infrastructure.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX][WIP] Initial version of diagnostics infrastructure."




[ghstack-poisoned]

BowenBao committed:
hack for sarif_om to pass linter; remove unused context on "[ONNX][WIP] Initial version of diagnostics infrastructure."




[ghstack-poisoned]

justinchuby commented on torch/onnx/diagnostic/__init__.py:
```diff
@@ -0,0 +1,14 @@
+from ._diagnostic import Diagnostic, Level, Rule
```
Should this be in `_internal`?

BowenBao replied:
Good idea, there has been some big changes in structure which will move a large portion of stuff to there. A quick question while we are on this. Is there a preference between `_internal._diagnostic/**.py` vs `_internal.diagnostic/**.py`? I noticed the `_beartype.py`  still has a underscore despite it being in internal folder already.

justinchuby replied:
_internal.diagnostic to me is preferable for its simplicity. _beartype was there to avoid redefining the beartype package's namespace

justinchuby commented on torch/onnx/diagnostic/formatter.py:
```diff
@@ -0,0 +1,32 @@
+            new_dict[new_k] = new_v
+        return new_dict
+
+    def from_camel_case_to_snake_case(s: str) -> str:
```
Mark: nested function def

justinchuby commented on torch/onnx/diagnostic/rules.yaml:
```diff
@@ -0,0 +1,81 @@
+# TODO: Separate rules into individual files.
+
+- id: RL0001
+  name: ONNXShapeInferenceIsMissingForNode
```
Any consideration for casing style from sarif? Have we considered snake_case? (More pythonic, aligns with mypy etc.)

BowenBao replied:
sarif format itself in json is camel cased, it is converted to snake case when generating the python object model. Yet this is somewhat in parallel to the format for actual name value, which is what I think your comment is referring to.

I like the dash style you are proposing, I'll push a change to see how it looks.

justinchuby commented on torch/onnx/diagnostic/rules.yaml:
```diff
@@ -0,0 +1,81 @@
+# Rules are based on sarif ReportingDescriptor format.
```
Love the yaml format

justinchuby commented on torch/onnx/diagnostic/rules.yaml:
```diff
@@ -0,0 +1,81 @@
+# TODO: Based on above, create helper script to generate new rules.
+# TODO: Separate rules into individual files.
+
+- id: RL0001
```
What is RL?

BowenBao replied:
Some random placeholder, updated to POE(PyTorch ONNX Exporter).

justinchuby commented on torch/onnx/errors.py:
```diff
@@ -5,7 +5,7 @@
 
 from torch import _C
-from torch.onnx import _constants
+from torch.onnx import _constants, diagnostic
```
`diagnostics`?

BowenBao replied:
I was going in circle about this as well for a while. I feel singular better depicts the system (e.g. DiagnosticEngine, DiagnosticSystem), while plural often represents actual "diagnostics". See https://github.dev/dotnet/roslyn/blob/main/src/Compilers/Core/Portable/Diagnostic/Diagnostic.cs .

justinchuby replied:
c# likes to name their files according to the class names, which imo is different than how we typically name python modules (usually a collection of types and functions). Happy to discuss further.

BowenBao replied:
Ah I should send a shorter link cause I was referring the folder name. But then I noticed the namespace is plural, hmm...

justinchuby replied:
😉

justinchuby commented on torch/onnx/sarif_om/__init__.py:
```diff
@@ -0,0 +1,65 @@
+# This file was generated by jschema_to_python version 1.2.3.
```
Hide this folder in _internal?

justinchuby commented on torch/onnx/utils.py:
```diff
@@ -175,6 +176,18 @@ def setup_onnx_logging(verbose: bool):
 
+@contextlib.contextmanager
+@_beartype.beartype
+def setup_onnx_diagnostic(verbose: bool):
```
`onnx_diagnostic_context`?

BowenBao committed:
Add back pass removed by accident; structural change; separate general diagnostic engine from export logic on "[ONNX][WIP] Initial version of diagnostics infrastructure."




[ghstack-poisoned]

BowenBao committed:
remove accidentally added file on "[ONNX][WIP] Initial version of diagnostics infrastructure."




[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX][WIP] Initial version of diagnostics infrastructure."




[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX][WIP] Initial version of diagnostics infrastructure."




[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Initial version of diagnostics infrastructure."




[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Initial version of diagnostics infrastructure."




[ghstack-poisoned]

justinchuby commented on test/onnx/test_diagnostics.py:
```diff
@@ -0,0 +1,239 @@
+
+def _assert_has_diagnostics(
+    engine: infra.DiagnosticEngine,
+    rule_level_pairs: Set[Tuple[infra.Rule, infra.Level]],
```
```suggestion
    rule_level_pairs: AbstractSet[Tuple[infra.Rule, infra.Level]],
```
https://gist.github.com/justinchuby/4021cebe9e093f636759a88de325c85f

BowenBao committed:
Add RuleCollection on "[ONNX] Initial version of diagnostics infrastructure."




[ghstack-poisoned]

BowenBao committed:
refine around Set on "[ONNX] Initial version of diagnostics infrastructure."




[ghstack-poisoned]

BowenBao committed:
fix copilot... on "[ONNX] Initial version of diagnostics infrastructure."




[ghstack-poisoned]

BowenBao committed:
small fix for start/end run; RuleCollection test on "[ONNX] Initial version of diagnostics infrastructure."




[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Initial version of diagnostics infrastructure."




[ghstack-poisoned]

BowenBao committed:
move setup_onnx_diagnostic to private on "[ONNX] Initial version of diagnostics infrastructure."




[ghstack-poisoned]

BowenBao committed:
Update docstring for engine on "[ONNX] Initial version of diagnostics infrastructure."




[ghstack-poisoned]

justinchuby commented on test/onnx/test_diagnostics.py:
```diff
@@ -0,0 +1,287 @@
+                    custom_rules.custom_rule_2, infra.Level.ERROR  # type: ignore[attr-defined]
+                )
+
+    def test_invalid_diagnostic_type_raises_type_error(self):
```
naming: suggested like above. here and elsewhere

justinchuby commented on torch/onnx/_internal/diagnostics/__init__.py:
```diff
@@ -0,0 +1,11 @@
+from ._diagnostic import context, engine, ExportDiagnostic, ExportDiagnosticTool, levels
```
levels: Can it be imported straight from infra?

justinchuby commented on torch/onnx/_internal/diagnostics/_diagnostic.py:
```diff
@@ -0,0 +1,92 @@
+    ) -> None:
+        super().__init__(rule, level, message_args, **kwargs)
+
+    def with_cpp_stack(self) -> "ExportDiagnostic":
```
`from typing_extension import Self`

```suggestion
    def with_cpp_stack(self) -> Self:
```

justinchuby commented on torch/onnx/utils.py:
```diff
@@ -172,6 +172,20 @@ def setup_onnx_logging(verbose: bool):
+    try:
+        yield diagnostics.context
+    finally:
+        diagnostics.context = _previous_context
```
What is the desired behavior for managing different contexts? When do we want to resume the previous context?

BowenBao replied:
Modified the code a bit wrt to variable names. I added docstring in `ExportDiagnosticEngine` for explanation. This function will likely move to the same file as `ExportDiagnosticEngine` per our other discussion.

justinchuby commented on torch/onnx/utils.py:
```diff
@@ -1869,7 +1902,8 @@ def register_custom_op_symbolic(
     versions = range(
-        max(_constants.ONNX_MIN_OPSET, opset_version), _constants.ONNX_MAX_OPSET + 1
+        max(_constants.ONNX_MIN_OPSET, opset_version),
+        _constants.ONNX_MAX_OPSET + 1,
```
I assume the changes where introduced either by a newer black version or because of the added final commas?

BowenBao replied:
good catch, I was unsure why this happened. Seems my local formatter added the comma.

thiagocrepaldi commented on torch/onnx/utils.py:
```diff
@@ -172,6 +172,20 @@ def setup_onnx_logging(verbose: bool):
 
+@contextlib.contextmanager
+@_beartype.beartype
+def _setup_onnx_diagnostic():
```
```suggestion
def _get_onnx_diagnostic():
```
`_setup_onnx_diagnostic() as diagnostic_ctx` feels a bit odd as `setup` routines usually do not return objects

justinchuby replied:
+1

justinchuby replied:
`_get_diagnostic_context` etc.

BowenBao replied:
how about `_create_diagnostic_context`?

justinchuby replied:
I like it

thiagocrepaldi commented on torch/onnx/_internal/diagnostics/infra/_infra.py:
```diff
@@ -0,0 +1,500 @@
+    """
+
+
+class DiagnosticContext:
```
Is this a context for "user" or for "onnx engineers"? can we have both?

Would be possible to have the engineer context to store the current method/function's input *automatically* captured  and stored as metadata?

For a crash on a symbolic, having the inputs from that particular symbolic function being saved for debugging would make debugging a lot easier.

justinchuby replied:
imo the context should contain all information possible. When we show it to users they get a view of what is appropriate.

Or maybe we allow different levels of detail?

BowenBao replied:
IMO context is bound to one particular export run. Through context api, engineers record diagnostics. Context is used inside the code block that needs diagnoses.

Engine provides the api to create contexts, as well as the api to inspect logs. So engine is used outside of the code block that needs diagnoses. If a users run a sequence of export calls in their script, each call will have a respectful 'run' log recorded in the final log retrievable through engine. 

For lightweight users, perhaps they don't need to interact with any of these. Maybe just an 'options' argument/environment variable to alter the verboseness or the level of details to (capture)display to them.

thiagocrepaldi replied:
I agree about having one context per run. My suggestion was that, in production, a "lightweight context" is used by default, just to give actionable error messages to the user, very succinct on what is displayed/stored.

Given the repro steps, an onnx engineer could enable the "debug context" and run the model again. On that second run, lots of more information would be presented/stored.

Maybe it is not the case of having different context, but using different `Level`(?) for the same "log"

BowenBao replied:
Agree, maybe one of, or both, of diagnostic options and environment variables (convenient for debugging without changing code).

BowenBao committed:
Migrate SARIF from attr to dataclasses with type annotation on "[ONNX] Initial version of diagnostics infrastructure."


Fix: https://github.com/pytorch/pytorch/issues/84659

[ghstack-poisoned]

BowenBao commented on test/onnx/test_diagnostics.py:
```diff
@@ -0,0 +1,287 @@
+
+
+@contextlib.contextmanager
+def assertAllDiagnostics(
```
I recall following the unittest convention, thoughts?

justinchuby replied:
I don't think we need to follow its convention - unit test was that way presumably because it was inspired from a Java library. New method we create can follow pep just fine

BowenBao commented on test/onnx/test_diagnostics.py:
```diff
@@ -0,0 +1,287 @@
+    def setUp(self):
+        engine = diagnostics.engine
+        engine.clear()
+        return super().setUp()
```
Why not? Is it that inherited test case don't need to call super setUp?

BowenBao commented on torch/csrc/onnx/diagnostics/diagnostics.h:
```diff
@@ -0,0 +1,68 @@
+static constexpr const char* const kPyLevelNames[] = {
+    "NONE",
+    "NOTE",
+    "WARNING",
```
iirc levels cannot be extended by sarif spec. We could use something like [tags](https://docs.oasis-open.org/sarif/sarif/v2.1.0/os/sarif-v2.1.0-os.html#_Toc34317450).

BowenBao replied:
Added comments for both python and C++. Mentioned tag but since there is no c++ api for it I also added a todo.

BowenBao commented on torch/onnx/_internal/diagnostics/__init__.py:
```diff
@@ -0,0 +1,11 @@
+from ._diagnostic import context, engine, ExportDiagnostic, ExportDiagnosticTool, levels
+from ._rules import rules
+
+__all__ = [
+    "ExportDiagnostic",
+    "ExportDiagnosticTool",
+    "rules",
+    "levels",
+    "engine",
+    "context",
+]
```
yea I believe it's a pytorch requirement.

BowenBao commented on torch/onnx/_internal/diagnostics/infra/_infra.py:
```diff
@@ -0,0 +1,500 @@
+"""This file defines an additional layer of abstraction on top of the SARIF OM."""
```
Mostly the former. There are a lots more features in sarif, the idea is to abstract away those that we need. Also we can have more customized components like specifying required fields or creating field mappings.

BowenBao committed:
Update on "[ONNX] Initial version of diagnostics infrastructure."


Fix: https://github.com/pytorch/pytorch/issues/84659

[ghstack-poisoned]

justinchuby commented on torch/onnx/utils.py:
```diff
@@ -174,16 +174,15 @@ def setup_onnx_logging(verbose: bool):
 @contextlib.contextmanager
 @_beartype.beartype
-def _setup_onnx_diagnostic():
+def _create_diagnostic_context():
```
Can this function live outside of utils and be moved to a relavant module in diagnostics?

BowenBao replied:
Let's keep it simple and put it in _diagnostics.py then. Renaming to `create_diagnostic_context_for_export` or `create_export_diagnostic_context`, wdyt?

justinchuby replied:
`create_export_diagnostic_context` sounds about right

BowenBao committed:
Update on "[ONNX] Initial version of diagnostics infrastructure."


Fix: https://github.com/pytorch/pytorch/issues/84659

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Initial version of diagnostics infrastructure."


Fix: https://github.com/pytorch/pytorch/issues/84659

[ghstack-poisoned]

BowenBao commented on torch/onnx/_internal/diagnostics/__init__.py:
```diff
@@ -0,0 +1,11 @@
+from ._diagnostic import context, engine, ExportDiagnostic, ExportDiagnosticTool, levels
```
These are great suggestions. Considering timeline of release I will try to create a follow up doc PR for cherry-picks.

BowenBao commented on torch/csrc/jit/passes/onnx/shape_type_inference.cpp:
```diff
@@ -1882,6 +1885,10 @@ void UpdateReliable(
         output->node()->kind().toDisplayString(),
         " type is missing, so it may result in wrong shape inference for the exported graph. ",
         "Please consider adding it in symbolic function.");
+    diagnostics::Diagnose(
```
> embed `TORCH_WARN`

Yes the actual emission can reuse whatever previous functionalities, potentially configurable too with DiagnosticOptions. Note that cpp api is supposed to be a thin wrapper over the actual library in Python, where the emission happens.

> optional `user_log`

This can be achieved with the `additional_message` argument. A bit more detail on how these info can be supplied. I created `diagostic.with_additional_message()` method to enable contiguous build up, but it makes it hard to decide emission time. Perhaps we'd still want a single constructor like function to record a diagnostic with all the info as arguments and be done with it? Let me know your thoughts.

> maps to a friendly message that is printed to stdout

Each rule does have several messages defined with them, they expose different level of details and information. The most informative one will also be used for generating the doc page of that rule under torch.onnx doc. You can view it at `rules.yaml`. (Will add proper documentation for these ... :))

thiagocrepaldi replied:
IIUC, `diagnostics::Diagnose` **already** calls `TORCH_WARN` through the use of `user_log` and `additional_message` and lines 1883-1887 could be removed?

Or that would be the behavior in a follow up PR?

BowenBao replied:
Right now `Diagnose` doesn't have any real user interacting behavior, i.e. it's not printing anything. Following up the emission of diagnosing will be added. When that happens, TORCH_WARN won't be needed. I'm adding comments above this to mention it's experimental and does not print anything yet.

BowenBao committed:
Update on "[ONNX] Initial version of diagnostics infrastructure."


Fix: https://github.com/pytorch/pytorch/issues/84659

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Initial version of diagnostics infrastructure."


~~Fix: https://github.com/pytorch/pytorch/issues/84659~~ There are unfinished work to close out the parent issue.
This PR introduces a general Python diagnostics infrastructure powered by SARIF,
and the exporter diagnostics module that builds on top of it.

[ghstack-poisoned]

BowenBao commented on torch/onnx/_internal/diagnostics/infra/engine.py:
```diff
@@ -0,0 +1,87 @@
+        >>> engine = DiagnosticEngine()
+
+        Step 4: Start a new diagnostic context.
+        >>> with engine.start_diagnostic_context(tool) as context:
```
```suggestion
        >>> with engine.create_diagnostic_context(tool) as context:
```

BowenBao committed:
Update torch/onnx/_internal/diagnostics/infra/engine.py

BowenBao committed:
Update on "[ONNX] Initial version of diagnostics infrastructure."


~~Fix: https://github.com/pytorch/pytorch/issues/84659~~ There are unfinished work to close out the parent issue.
This PR introduces a general Python diagnostics infrastructure powered by SARIF,
and the exporter diagnostics module that builds on top of it.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Initial version of diagnostics infrastructure."


~~Fix: https://github.com/pytorch/pytorch/issues/84659~~ There are unfinished work to close out the parent issue.
This PR introduces a general Python diagnostics infrastructure powered by SARIF,
and the exporter diagnostics module that builds on top of it.

[ghstack-poisoned]

BowenBao commented on torch/csrc/onnx/diagnostics/generated/rules.h:
```diff
@@ -0,0 +1,47 @@
+/**
+ * Diagnostic rules for PyTorch ONNX export.
+ *
+ * This file is generated by gen_diagnostics.py. Do not edit directly.
```
done

justinchuby commented on test/onnx/internal/test_diagnostics.py:
```diff
@@ -0,0 +1,285 @@
+    engine: infra.DiagnosticEngine,
+    rule_level_pairs: AbstractSet[Tuple[infra.Rule, infra.Level]],
+):
+    """
+    Context manager to assert that all diagnostics are emitted.
+
+    Usage:
```
Very minor; doesn't need changes, just to point it out

```suggestion
    """Context manager to assert that all diagnostics are emitted.

    Usage:
```

justinchuby commented on test/onnx/internal/test_diagnostics.py:
```diff
@@ -0,0 +1,285 @@
+
+
+def assert_diagnostic(
+    test_suite: common_utils.TestCase,
```
```suggestion
    test_suite: unittest.TestCase,
```

nit, optional: Can it work with arbitrary unittest test cases?

BowenBao replied:
I think so, updated.

justinchuby commented on test/onnx/internal/test_diagnostics.py:
```diff
@@ -0,0 +1,285 @@
+    return assert_all_diagnostics(test_suite, engine, {(rule, level)})
+
+
+class TestONNXDiagnostics(common_utils.TestCase):
```
```suggestion
class TestOnnxDiagnostics(common_utils.TestCase):
```

nit

justinchuby commented on test/onnx/internal/test_diagnostics.py:
```diff
@@ -0,0 +1,285 @@
+    ):
+        with self.assertRaisesRegex(
+            TypeError,
+            "Expected diagnostic_type to be a subclass of Diagnostic, " "but got",
```
```suggestion
            "Expected diagnostic_type to be a subclass of Diagnostic, but got",
```
nit

justinchuby commented on tools/onnx/gen_diagnostics.py:
```diff
@@ -0,0 +1,185 @@
+    return name.replace("-", "_")
+
+
+def _kebab_case_to_pascal_case(name: str) -> str:
```
🤣

BowenBao replied:
🥷 🥷

justinchuby commented on tools/onnx/gen_diagnostics.sh:
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+cd ../../
```
Probably make this more robust?

BowenBao replied:
done, added a few lines of comments too.

justinchuby commented on torch/csrc/onnx/diagnostics/diagnostics.h:
```diff
@@ -0,0 +1,71 @@
+ * modifiable. For alternative categories, please use Tag instead.
+ * @todo Introduce Tag to C++ api.
+ */
+enum class Level : uint32_t {
```
Just for my learning, no changes needed: do we need uint32 or can a smaller type do it?

BowenBao replied:
Good point, I think `uint8_t` will do.

justinchuby commented on torch/onnx/_internal/diagnostics/_diagnostic.py:
```diff
@@ -0,0 +1,121 @@
+    def with_cpp_stack(self: _ExportDiagnostic) -> _ExportDiagnostic:
+        # TODO: Implement this.
+        # self.stacks.append(...)
+        return self
```
nit: optional: raise unimplemented error?

justinchuby commented on torch/onnx/_internal/diagnostics/infra/_infra.py:
```diff
@@ -0,0 +1,349 @@
+
+
+@dataclasses.dataclass(frozen=True)
+class DiagnosticTool:
```
Is this class mainly for carrying data or does it do something else? For classes whose main job is more than storing data I'd suggest not using a dataclass

BowenBao replied:
Good question. It is not decided. So far other than storing stuff, it acts as a relay with a few checks. I'll leave this as is for now.

justinchuby commented on torch/onnx/_internal/diagnostics/infra/_infra.py:
```diff
@@ -0,0 +1,349 @@
+            new_collection_class_name,
+            [
+                (
+                    formatter._kebab_case_to_snake_case(rule.name),
```
Looks like `_kebab_case_to_snake_case` is used outside of `formatter`. I would rename it to `kebab_case_to_snake_case` to make it public

justinchuby commented on torch/onnx/_internal/diagnostics/infra/formatter.py:
```diff
@@ -0,0 +1,47 @@
+    return re.sub(r"([A-Z])", r"_\1", s).lower()
+
+
+def _kebab_case_to_snake_case(s: str) -> str:
```
```suggestion
def kebab_case_to_snake_case(s: str) -> str:
```

justinchuby commented on torch/onnx/_internal/diagnostics/infra/formatter.py:
```diff
@@ -0,0 +1,47 @@
+
+def _convert_key(
+    object: Union[Dict[str, Any], Any], convert: Callable[[str], str]
+) -> Union[Dict[str, Any], Any]:
```
To me what this function does is not very clear (or why/when one needs to use it). Consider a brief docstring?

justinchuby replied:
I can see now from the usage in `sarif_to_json`. Still, just a sentence should suffice.

BowenBao replied:
added some docstring

justinchuby commented on torch/onnx/_internal/diagnostics/infra/formatter.py:
```diff
@@ -0,0 +1,47 @@
+
+
+def _convert_key(
+    object: Union[Dict[str, Any], Any], convert: Callable[[str], str]
+) -> Union[Dict[str, Any], Any]:
```
```suggestion
    object: Union[Dict[str, _T1], _T2], convert: Callable[[str], str]
) -> Union[Dict[str, _T1], _T2]:
```

nit: optional

justinchuby replied:
ok!

justinchuby commented on torch/onnx/_internal/diagnostics/infra/sarif/version.py:
```diff
@@ -1,3 +1,5 @@
-SARIF_SCHEMA_LINK = "https://docs.oasis-open.org/sarif/sarif/v2.1.0/cs01/schemas/sarif-schema-2.1.0.json"
+from typing_extensions import Final
+
+SARIF_VERSION: Final = "2.1.0"
```
Final: nice: learned something new!

BowenBao replied:
Adding type annotation to sarif is very helpful. I found myself checking the original spec much less now that I could jump around instead. And this `Final` was forced upon after the version attribute in sarif was annotated as Literal[...].

justinchuby commented on torch/csrc/onnx/diagnostics/generated/rules.h:
```diff
@@ -0,0 +1,48 @@
+    "node_missing_onnx_shape_inference",
+    "missing_custom_symbolic_function",
+    "missing_standard_symbolic_function",
+    "operator_supported_in_newer_opset_version",
```
Do we need to repeat the names of the rules here?

BowenBao replied:
Haven't found a good way for c++ to turn enum to string other than this, so ...

justinchuby commented on torch/onnx/errors.py:
```diff
@@ -54,15 +59,38 @@ def __init__(
+                diagnostics.rules.missing_custom_symbolic_function,
+                diagnostics.levels.ERROR,
+                message_args=(f"{domain}::{op_name}",),
+            )
```
Would it be possible to get the message (for the python error) from the rule in the future?

BowenBao replied:
Tricky to have a good ux experience with message and arguments. Getting message alone is simple, but it's one step(f12) away from knowing what exact arguments it's expecting. 
Tried a few approaches locally. The best ux friendly one is generating individual class for each rule with "overloaded" arguments. LMK if you have ideas...

justinchuby replied:
Thanks for the explanation! I will think about it

justinchuby reviewed (approved):
LGTM🚀

justinchuby commented:
![image](https://user-images.githubusercontent.com/11205048/193166649-b1a75662-f44f-45bd-b4e3-e0c5823fdc8a.png)

fyi this is likely still going to close the issue.

BowenBao committed:
Update on "[ONNX] Initial version of diagnostics infrastructure."


~~Fix: https://github.com/pytorch/pytorch/issues/84659~~ There are unfinished work to close out the parent issue.
This PR introduces a general Python diagnostics infrastructure powered by SARIF,
and the exporter diagnostics module that builds on top of it.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Initial version of diagnostics infrastructure."


~~Fix: https://github.com/pytorch/pytorch/issues/84659~~ There are unfinished work to close out the parent issue.
This PR introduces a general Python diagnostics infrastructure powered by SARIF,
and the exporter diagnostics module that builds on top of it.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Initial version of diagnostics infrastructure."


This PR introduces a general Python diagnostics infrastructure powered by SARIF,
and the exporter diagnostics module that builds on top of it.

[ghstack-poisoned]

BowenBao commented:
@pytorchbot merge -g

justinchuby commented:
Need to start merge from https://github.com/pytorch/pytorch/pull/85428 or no?

BowenBao commented:
> Need to start merging from #85428 or no?

I thought stack merge works hmm

BowenBao commented:
@pytorchbot merge -g

BowenBao commented:
@pytorchbot merge -g

BowenBao commented:
Thanks @justinchuby @AllenTiTaiWang @thiagocrepaldi @abock for the detailed reviews!
